src/include/OSL/dual.h

```diff
@@ -287,533 +404,588 @@ inline Dual2<T> operator/ (const Dual2<T> &a, const T &b)
 {
-    T expa = std::exp(a.val());
-    return Dual2<T> (expa, expa * a.dx(), expa * a.dy());
+    T f = std::exp(a.val());
+    return dualfunc (a, f, f);
 }
 
-inline Dual2<float> fast_exp(const Dual2<float> &a)
+template<class T, int P>
+inline Dual<T,P> fast_exp(const Dual<T,P> &a)
 {
-    float expa = OIIO::fast_exp(a.val());
-    return Dual2<float> (expa, expa * a.dx(), expa * a.dy());
+    T f = OIIO::fast_exp(a.val());
+    return dualfunc (a, f, f);
 }
 
 // f(x) = 2^x, f'(x) = (2^x)*log(2)
-template<class T>
-inline Dual2<T> exp2 (const Dual2<T> &a)
+template<class T, int P>
+inline Dual<T,P> exp2 (const Dual<T,P> &a)
 {
     // FIXME: std::exp2 is only available in C++11
-    T exp2a = exp2f(float(a.val()));
-    return Dual2<T> (exp2a, exp2a*T(M_LN2)*a.dx(), exp2a*T(M_LN2)*a.dy());
+    T f = exp2f(float(a.val()));
+    return dualfunc (a, f, f*T(M_LN2));
 }
 
-inline Dual2<float> fast_exp2(const Dual2<float> &a)
+template<class T, int P>
+inline Dual<T,P> fast_exp2(const Dual<T,P> &a)
 {
-    float exp2a = OIIO::fast_exp2(float(a.val()));
-    return Dual2<float> (exp2a, exp2a*float(M_LN2)*a.dx(), exp2a*float(M_LN2)*a.dy());
+    T f = OIIO::fast_exp2(float(a.val()));
+    return dualfunc (a, f, f*T(M_LN2));
 }
 
 
 // f(x) = e^x - 1, f'(x) = e^x
-template<class T>
-inline Dual2<T> expm1 (const Dual2<T> &a)
+template<class T, int P>
+inline Dual<T,P> expm1 (const Dual<T,P> &a)
 {
     // FIXME: std::expm1 is only available in C++11
-    T expm1a = expm1f(float(a.val())); // float version!
-    T expa   = std::exp  (a.val());
-    return Dual2<T> (expm1a, expa * a.dx(), expa * a.dy());
+    T f  = expm1f(float(a.val())); // float version!
+    T df = std::exp  (a.val());
+    return dualfunc (a, f, df);
 }
 
-inline Dual2<float> fast_expm1(const Dual2<float> &a)
+template<class T, int P>
+inline Dual<T,P> fast_expm1(const Dual<T,P> &a)
 {
-    float expm1a = OIIO::fast_expm1(a.val());
-    float expa   = OIIO::fast_exp  (a.val());
-    return Dual2<float> (expm1a, expa * a.dx(), expa * a.dy());
+    T f  = OIIO::fast_expm1(a.val());
+    T df = OIIO::fast_exp  (a.val());
+    return dualfunc (a, f, df);
 }
 
 // f(x) = erf(x), f'(x) = (2e^(-x^2))/sqrt(pi)
-template<class T>
-inline Dual2<T> erf (const Dual2<T> &a)
+template<class T, int P>
+inline Dual<T,P> erf (const Dual<T,P> &a)
 {
     // FIXME: std::erf is only defined in C++11
-    T erfa = erff (float(a.val())); // float version!
-    T two_over_sqrt_pi = T(1.128379167095512573896158903);
-    T derfadx = std::exp (-a.val() * a.val()) * two_over_sqrt_pi;
-    return Dual2<T> (erfa, derfadx * a.dx(), derfadx * a.dy());
+    T f = erff (float(a.val())); // float version!
+    const T two_over_sqrt_pi = T(1.128379167095512573896158903);
+    T df = std::exp (-a.val() * a.val()) * two_over_sqrt_pi;
+    return dualfunc (a, f, df);
 }
 
-inline Dual2<float> fast_erf(const Dual2<float> &a)
+template<class T, int P>
+inline Dual<T,P> fast_erf(const Dual<T,P> &a)
```

Comment:
The old implementation only made the `fast_*` calls available for `T=float`. Do you think there is any value in supporting the fast implementations for doubles given they are all float internally?

Reply:
I was trying to specifically avoid hard-coding anything to be float only, unless I truly thought the only possible implementation was for single, scalar floats.

I want to keep all the Dual functionality maximally composable with the SIMD types (we allow `fast_exp(vfloatN)`, so why not allow a version with derivatives without any additional code?) as well as Imath types (if we add a `fast_exp(Vec3)` that operates per-component, this template would also give us a `fast_exp(Dual<Vec3>)`).

Using it with double will already be a build-time error, since no fast_blah(double) will be found most of the time.

Reply:
Ah I wasn't thinking of the SIMD types. Make sense as it is then.

`fast_blah(double)` doesn't exist but float is convertible to `double` so I think the compiler will accept it anyhow.